## The goal of this assigment is to discover what the best date night movie is through the sorting of our data. Below we import our libraries and set our directory in order to get the data.

In [2]:
import os 
os.chdir("/Users/JackVoigt/Documents Local/MSIA Files/CSC 605/CSC_605_Repo/Assignment/Assignment_2/data")
import pandas as pd
import numpy as np

## Next we can read in our datasets that contain info for users, ratings, and movies. We give each of the .dat files the names for the columns and read in the data afterwards, assigning the names.

In [8]:
unames = ['user_id','gender','age','occupation','zip']

In [9]:
users = pd.read_table(os.path.join('users.dat'), 
                          sep='::', header=None, names=unames)

/Users/JackVoigt/opt/anaconda3/lib/python3.8/site-packages/pandas/io/parsers.py:765: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  return read_csv(**locals())


In [27]:
users.head(10)

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455
5        6      F   50           9  55117
6        7      M   35           1  06810
7        8      M   25          12  11413
8        9      M   25          17  61614
9       10      F   35           1  95370

In [30]:
users.shape

(6040, 5)

In [12]:
rnames = ['user_id', 'movie_id', 'rating', 'timestamp']

In [13]:
ratings = pd.read_table(os.path.join('ratings.dat'), 
                            sep='::', header=None, names=rnames)

/Users/JackVoigt/opt/anaconda3/lib/python3.8/site-packages/pandas/io/parsers.py:765: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  return read_csv(**locals())


In [28]:
ratings.head(10)

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291
5        1      1197       3  978302268
6        1      1287       5  978302039
7        1      2804       5  978300719
8        1       594       4  978302268
9        1       919       4  978301368

In [31]:
ratings.shape

(1000209, 4)

In [17]:
mnames = ['movie_id', 'title','genres']

In [18]:
movies = pd.read_table(os.path.join('movies.dat'), 
                           sep='::', header=None, names=mnames)

/Users/JackVoigt/opt/anaconda3/lib/python3.8/site-packages/pandas/io/parsers.py:765: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  return read_csv(**locals())


In [29]:
movies.head(10)

movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   
5         6                         Heat (1995)         Action|Crime|Thriller   
6         7                      Sabrina (1995)                Comedy|Romance   
7         8                 Tom and Huck (1995)          Adventure|Children's   
8         9                 Sudden Death (1995)                        Action   
9        10                    GoldenEye (1995)     Action|Adventure|Thriller   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II  
5  1995                         Heat  
6  1995                      Sabrina  
7  1995                 Tom and Huck  
8  1995                 Sudden Death  
9  1995                    GoldenEye

In [32]:
movies.shape

(3883, 5)

## Next, we can clean up our movie titles

In [21]:
tmp = movies.title.str.extract('(.*) \(([0-9]+)\)')

In [22]:
tmp.apply(lambda x:x[0] if len(x) > 0 else None)

0    Toy Story
1         1995
dtype: object

In [23]:
tmp.apply(lambda x: x[0][:40] if len(x) > 0 else None)

0    Toy Story
1         1995
dtype: object

In [24]:
movies['year'] = tmp[1]

In [25]:
movies['short_title'] = tmp[0]

In [26]:
movies.head()

movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II

## Now we have created a new column titled "short title" that doesnt include the year. Now, we can look to merge our datasets with the pandas merge function. We can start by inner joining our tables users and ratings on the user id column. This will merge the datasets that have a one to many relationship within this column, and an inner join should merge these datasets on matching values.

In [33]:
full_df = pd.merge(users, ratings, on = "user_id")

In [35]:
full_df.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp
0        1      F    1          10  48067      1193       5  978300760
1        1      F    1          10  48067       661       3  978302109
2        1      F    1          10  48067       914       3  978301968
3        1      F    1          10  48067      3408       4  978300275
4        1      F    1          10  48067      2355       5  978824291

In [36]:
full_df.shape

(1000209, 8)

## The first merge is shown above, as we can see we retain the same amount of observations but at 4 new columns. We can now merge this new dataset with the dataset "movies", but this time on the column "movie_id".

In [40]:
full_df = pd.merge(full_df, movies, on = "movie_id")

In [41]:
full_df.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  year  \
0  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
1  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
2  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
3  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
4  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   

                       short_title  
0  One Flew Over the Cuckoo's Nest  
1  One Flew Over the Cuckoo's Nest  
2  One Flew Over the Cuckoo's Nest  
3  One Flew Over the Cuckoo's Nest  
4  One Flew Over the Cuckoo's Nest

In [42]:
full_df.shape

(1000209, 12)

## As expected, we can merge our datasets to now add 4 new columns to our dataset. Our new dataset, "full_df" includes the columns from all 3 of the datasets we read in, but the same amount of observations as we had ratings with 1,000,209. Now, we can find the highest rated movie, as well as the best movie for date night based on some condition.

In [107]:
ratings = full_df.groupby('movie_id', as_index=False)['rating'].mean()

In [108]:
ratings

movie_id    rating
0            1  4.146846
1            2  3.201141
2            3  3.016736
3            4  2.729412
4            5  3.006757
...        ...       ...
3701      3948  3.635731
3702      3949  4.115132
3703      3950  3.666667
3704      3951  3.900000
3705      3952  3.780928

[3706 rows x 2 columns]

In [109]:
ratings = ratings.sort_values(by="rating", ascending = False)

In [110]:
ratings.head(20)

movie_id    rating
926        989  5.000000
3635      3881  5.000000
1652      1830  5.000000
3152      3382  5.000000
744        787  5.000000
3054      3280  5.000000
3367      3607  5.000000
3010      3233  5.000000
2955      3172  5.000000
3414      3656  5.000000
3021      3245  4.800000
51          53  4.750000
2309      2503  4.666667
2698      2905  4.608696
1839      2019  4.560510
309        318  4.554558
802        858  4.524966
708        745  4.520548
49          50  4.517106
513        527  4.510417

## Above we have sorted all of the movies by the mean of all of their ratings, which can be identified by thier movie_id. We can see a handful of movies that have recieved 5 stars, any of which can be found by name, such as the top movie with an id of 926. 

In [111]:
full_df[full_df["movie_id"] == 926]

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
487124       10      F   35           1  95370       926       4  978226212   
487125       27      M   25          11  19130       926       5  978130018   
487126       33      M   45           3  55421       926       3  978982783   
487127       48      M   25           4  92107       926       4  978065606   
487128       58      M   25           2  30303       926       4  977937340   
...         ...    ...  ...         ...    ...       ...     ...        ...   
487522     5974      F   25           1  32303       926       5  956947169   
487523     5986      F   56           1  01060       926       4  956887137   
487524     6016      M   45           1  37209       926       3  956957670   
487525     6037      F   45           1  76006       926       5  956709701   
487526     6039      F   45           0  01060       926       5  956706324   

                       title genres  year    short_title  
487124  All About Eve (1950)  Drama  1950  All About Eve  
487125  All About Eve (1950)  Drama  1950  All About Eve  
487126  All About Eve (1950)  Drama  1950  All About Eve  
487127  All About Eve (1950)  Drama  1950  All About Eve  
487128  All About Eve (1950)  Drama  1950  All About Eve  
...                      ...    ...   ...            ...  
487522  All About Eve (1950)  Drama  1950  All About Eve  
487523  All About Eve (1950)  Drama  1950  All About Eve  
487524  All About Eve (1950)  Drama  1950  All About Eve  
487525  All About Eve (1950)  Drama  1950  All About Eve  
487526  All About Eve (1950)  Drama  1950  All About Eve  

[403 rows x 12 columns]

## We can see the top 5 star movie is titled "All About Eve" and we could do this for any of the highest rated movies in our dataset.

## Now we will look at finding a good rated movie for date night based on whatever criteria we so choose. First, we want to find all of the movies that were voted at least 4 stars on average by both men and women. Then we can filter out certain genres that we feel are better suited for a date night.

In [112]:
male_df = full_df[full_df["gender"] == "M"]

In [113]:
male_df.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   
6       19      M    1          10  48073      1193       5  982730936   

                                    title genres  year  \
1  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
2  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
3  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
4  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
6  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   

                       short_title  
1  One Flew Over the Cuckoo's Nest  
2  One Flew Over the Cuckoo's Nest  
3  One Flew Over the Cuckoo's Nest  
4  One Flew Over the Cuckoo's Nest  
6  One Flew Over the Cuckoo's Nest

In [114]:
female_df = full_df[full_df["gender"] == "F"]

In [115]:
female_df.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0         1      F    1          10  48067      1193       5  978300760   
5        18      F   18           3  95825      1193       4  978156168   
7        24      F   25           7  10023      1193       5  978136709   
8        28      F   25           1  14607      1193       3  978125194   
19       59      F   50           1  55413      1193       4  977934292   

                                     title genres  year  \
0   One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
5   One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
7   One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
8   One Flew Over the Cuckoo's Nest (1975)  Drama  1975   
19  One Flew Over the Cuckoo's Nest (1975)  Drama  1975   

                        short_title  
0   One Flew Over the Cuckoo's Nest  
5   One Flew Over the Cuckoo's Nest  
7   One Flew Over the Cuckoo's Nest  
8   One Flew Over the Cuckoo's Nest  
19  One Flew Over the Cuckoo's Nest

In [116]:
male_ratings = male_df.groupby('movie_id', as_index=False)['rating'].mean()

In [117]:
male_ratings = male_ratings.sort_values(by="rating", ascending = False)

In [124]:
male_ratings.head(20)

movie_id    rating
737        787  5.000000
913        985  5.000000
2985      3233  5.000000
3029      3280  5.000000
2931      3172  5.000000
420        439  5.000000
124        130  5.000000
3381      3656  5.000000
1637      1830  5.000000
917        989  5.000000
3254      3517  5.000000
2700      2931  4.833333
2996      3245  4.750000
579        598  4.666667
51          53  4.666667
2675      2905  4.639344
2287      2503  4.600000
3083      3338  4.583333
795        858  4.583333
1822      2019  4.576628

In [103]:
female_ratings = female_df.groupby('movie_id', as_index=False)['rating'].mean()

In [119]:
female_ratings = female_ratings.sort_values(by="rating", ascending = False)

In [126]:
female_ratings.head(28)

movie_id    rating
50          53  5.000000
3356      3817  5.000000
1227      1420  5.000000
2007      2332  5.000000
1829      2129  5.000000
2889      3292  5.000000
2645      3012  5.000000
1251      1450  5.000000
2118      2444  5.000000
2616      2981  5.000000
2208      2538  5.000000
697        787  5.000000
2961      3382  5.000000
929       1071  5.000000
3197      3641  5.000000
2330      2675  5.000000
1511      1780  5.000000
3165      3607  5.000000
2410      2760  5.000000
377        394  5.000000
627        687  5.000000
2847      3245  5.000000
624        681  5.000000
742        854  5.000000
3420      3888  5.000000
3413      3881  5.000000
988       1153  5.000000
618        670  4.842105

## We can see there are 11 movies that males have at a perfect 5 star rating and 27 movies that females have at a perfect 5 star rating. we can slice these series so only the 5 star movies remain from the list of descending rankings.

In [127]:
male_ratings = male_ratings[:11]

In [129]:
male_ratings.head(20)

movie_id  rating
737        787     5.0
913        985     5.0
2985      3233     5.0
3029      3280     5.0
2931      3172     5.0
420        439     5.0
124        130     5.0
3381      3656     5.0
1637      1830     5.0
917        989     5.0
3254      3517     5.0

In [130]:
female_ratings = female_ratings[:27]

In [132]:
female_ratings.head(30)

movie_id  rating
50          53     5.0
3356      3817     5.0
1227      1420     5.0
2007      2332     5.0
1829      2129     5.0
2889      3292     5.0
2645      3012     5.0
1251      1450     5.0
2118      2444     5.0
2616      2981     5.0
2208      2538     5.0
697        787     5.0
2961      3382     5.0
929       1071     5.0
3197      3641     5.0
2330      2675     5.0
1511      1780     5.0
3165      3607     5.0
2410      2760     5.0
377        394     5.0
627        687     5.0
2847      3245     5.0
624        681     5.0
742        854     5.0
3420      3888     5.0
3413      3881     5.0
988       1153     5.0

In [133]:
combined_ratings = pd.merge(male_ratings, female_ratings, on = "movie_id")

In [134]:
combined_ratings.head()

movie_id  rating_x  rating_y
0       787       5.0       5.0

## We can see there is only one movie that has a 5 star rating for both men and women. The title of this movie can be found by locating the movie that has a movie id of 787.

In [135]:
full_df[full_df["movie_id"] == 787]

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
965717      149      M   25           1  29205       787       5  977325719   
965718     2825      F   25          20  94014       787       5  972610193   
965719     2872      M   25          20  94014       787       5  972423586   

                                     title       genres  year  \
965717  Gate of Heavenly Peace, The (1995)  Documentary  1995   
965718  Gate of Heavenly Peace, The (1995)  Documentary  1995   
965719  Gate of Heavenly Peace, The (1995)  Documentary  1995   

                        short_title  
965717  Gate of Heavenly Peace, The  
965718  Gate of Heavenly Peace, The  
965719  Gate of Heavenly Peace, The

## The only movie that has a mean rating of 5 stars for both men and women is "The Gate of Heavenly Peace", which makes it the perfect movie for date night.